core/src/mender-alloc.c

```diff
@@ -0,0 +1,68 @@
+/**
+ * @file      mender-alloc.c
+ * @brief     Platform-independent parts of the Mender memory management implementation
+ *
+ * Copyright Northern.tech AS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdlib.h> /* malloc(),... */
+#include <string.h> /* memset() */
+
+#include "mender-alloc.h"
+
+static MenderAllocator   malloc_fn  = NULL;
+static MenderReallocator realloc_fn = NULL;
+static MenderDeallocator free_fn    = NULL;
+
+void
+mender_set_allocation_funcs(MenderAllocator malloc_func, MenderReallocator realloc_func, MenderDeallocator free_func) {
+    malloc_fn  = malloc_func;
+    realloc_fn = realloc_func;
+    free_fn    = free_func;
```

Comment:
Should we allow passing only one of the three?

As it is now because we just check individually if they are NULL when using them (ie mender_malloc checks or calls malloc) we could end-up with a user specifying only one and having really bad memory allocation (like custom malloc and libc realloc, or custom malloc and no custom free :lady_beetle: )

I suggest we assert on this - it should be caught by a developer at dev time

**Edit:** I see now that this is handled in mender-client code. However, with the future vision of using the project also as an SDK I think we should move the checks here.

Reply:
> However, with the future vision of using the project also as an SDK I think we should move the checks here.

In that future version, I don't want this function to be part of the SDK/public API. The only way to set the allocation functions should be via `mender_client_init()`. Otherwise we ask for trouble because calling this function at some other point would cause issues too.

Reply:
Okey, makes sense. Thank you for the reasoning.